### PR TITLE
Re-init for paused multi-bitrate videos when bitrate changes.

### PR DESCRIPTION
**Pull request by cramforce**
Besides this feature, this PR also comes with a minimal `WeakRef` special-case-partial-fill for DOM Elements. @jridgewell may be interested in reviewing.

Part of #29254

### CODE REVIEW

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 1]
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+import {DomBasedWeakRef} from '../../../src/utils/dom-based-weakref';
```
Comment (jridgewell) [inline extensions/amp-video/0.1/flexible-bitrate.js:17]: Did you forget to commit this file?

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 2]
```diff
@@ -213,6 +219,26 @@ export class BitrateManager {
...
     });
   }
+
+  /**
+   * Update other managed videos when we learn that the current selected
+   * bandwidth wasn't good. Only operates on videos that are currently paused
+   * as we never want to interrupt playing videos if we don't have to.
+   */
+  updateOtherManagedAndPausedVideos_() {
+    this.videos_.forEach((weakref, index) => {
```
Comment (jridgewell) [inline extensions/amp-video/0.1/flexible-bitrate.js:229]: Doing a forEach while mutating will cause index skips. This needs to be a manual for-loop with a `i--` to guarantee every element is checked.

```js
a = [1, 2, 3];
a.forEach((e, i) => {
    if (e >= 2) a.splice(i, 1);
});
assert.deepEqual(a, [1, 3]);
```

File: src/utils/dom-based-weakref.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * Returns a WeakRef. Uses this implementation if the real WeakRef class
+   * is not available.
+   * @param {!Window} win
+   * @param {!Element} element
+   * @return {!WeakRef<!Element>|DomBasedWeakRef<!Element}
+   */
+  static make(win, element) {
+    if (win.WeakRef) {
+      return new WeakRef(element);
```
Comment (jridgewell) [inline src/utils/dom-based-weakref.js:46]: Nit: 
```suggestion
      return new win.WeakRef(element);
```

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 1 above)
Comment (cramforce) [reply extensions/amp-video/0.1/flexible-bitrate.js:17]: I had just noticed. Fixed

File: src/utils/dom-based-weakref.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    /** @private @const */
+    this.id_ = id;
+  }
+
+  /**
+   * Returns a WeakRef. Uses this implementation if the real WeakRef class
+   * is not available.
+   * @param {!Window} win
+   * @param {!Element} element
+   * @return {!WeakRef<!Element>|DomBasedWeakRef<!Element}
```
Comment (jridgewell) [inline src/utils/dom-based-weakref.js:42]:
```suggestion
   * @return {!WeakRef<!Element>|DomBasedWeakRef<!Element>}
```

File: src/utils/dom-based-weakref.js (hunk 3 above)
Comment (cramforce) [reply src/utils/dom-based-weakref.js]: Done

File: extensions/amp-video/0.1/flexible-bitrate.js (hunk 2 above)
Comment (cramforce) [reply extensions/amp-video/0.1/flexible-bitrate.js]: Ha, I vaguely recalled this, but my too naive testing failed to detect this :)

File: src/utils/dom-based-weakref.js (hunk 4 above)
Comment (cramforce) [reply src/utils/dom-based-weakref.js]: Done.

I just fixed the typing. Had to add a temporary extern for `WeakRef`

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 5]
```diff
@@ -213,6 +219,27 @@ export class BitrateManager {
       dev().fine(TAG, 'Playing at lower bitrate %s', video.currentSrc);
     });
   }
+
+  /**
+   * Update other managed videos when we learn that the current selected
+   * bandwidth wasn't good. Only operates on videos that are currently paused
+   * as we never want to interrupt playing videos if we don't have to.
+   */
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:227]:
```suggestion
   * @private
   */
```

**Comment by gmajoulet** [comment]
I'll file a ticket so we do follow up work to optimize this. The biggest thing on my mind is that if there are many other managed videos, downgrading the bitrate and trying to (pre)load all of them at once will maybe make things even worse for the playback of the playing video - I'm concerned about both hardware decoding and network. Stories could set some attribute on videos that it wants to downgrade.

**Comment by cramforce** [comment]
Yeah, I've been wondering on the "storm of preloads" as well. The media pool should probably handle it by managing where to apply a preload attribute. Or the multi-bitrate code should delete the preload to be cautious.

**Comment by gmajoulet** [comment]
I'm tracking follow up improvements here: https://github.com/ampproject/amphtml/projects/111

**Pull request merged**